### DISCUSSION TITLE
Make BBOT compatible with Ax/Axiom

### DISCUSSION THREAD

**Discussion by Sh4d0wHunt3rX** [blacklanternsecurity/bbot#1145] [Ideas]
When I run BBOT, I need a very powerful server, but when it's not in process, I need a very cheap server to manage the data. I was thinking that would be awesome if possible to make BBOT compatible with AXIOM, so when we run it, it creates a powerful server and when it is finished, terminate that server and (move data to my main server?). Not sure if there is any other solution for this, axiom came to my mind. Maybe kubernete? (I have never worked with it)

https://github.com/pry0cc/axiom

**Comment by TheTechromancer** [comment]
This is a good idea. I'll look into axiom.

**Comment by Sh4d0wHunt3rX** [comment]
Axiom also can make processes parallel. For example, paramminer is a slow module, so we run it on 5 instances at the same time to finish it sooner. Maybe possible to make something like this as built-in in bbot. Same as "file splitter" in Trickest. In trickest, if we put file splitter before of any module, then that module will run on several machines concurrently. So, maybe the whole bbot can have axiom compatibility or some slow modules can run on several instances at the same time.

![image](https://github.com/blacklanternsecurity/bbot/assets/15929497/a56885e1-a7d2-4017-b541-bd5ce059c393)

Thanks 🙏

**Reply by TheTechromancer** [reply]
BBOT already does this too. By default the paramminer modules run in parallel 12 instances at a time.

You can customize it through the config:

```bash
bbot -c modules.paramminer_headers.max_event_handlers=20
```

**Comment by Sh4d0wHunt3rX** [comment]
Or maybe shadowclone
https://github.com/fyoorer/ShadowClone

This actually can be used also as WAF bypass method

**Comment by Sh4d0wHunt3rX** [comment]
The new tool in this category is
https://github.com/defparam/lemma

If something like iis scanner, bypass 403, ffuf ... can execute at the end as serverless functions, can reduce the bbot scanning time significantly.

**Comment by Sh4d0wHunt3rX** [comment]
The defparam version is 404, but there are some forks available. Recently I have disabled some bbot modules such as shortscan or ffuf or bypass403 and doing these in lemma/shadowclone , I know it's not recursive on my side anymore but the reduction in scan time is significant. 

I was thinking if possible to integrate this ability as an option in bbot, that possible to enable/disable it and do some of the operations in aws lambda, bbot will go in another level. Imagine, ffuf, bypass403, shortscan ... in 1000 parallel executions, how much fast can bbot be.

**Comment by bc0la** [comment]
Being able to distribute things like port scans/ffuf/gowitness/nuclei with axiom would be so awesome. It's probably the one thing holding us back from using bbot full time. (particularly on large scopes)

**Reply by TheTechromancer** [reply]
I agree this would be a good feature to have. I foresee BBOT having an option like this in the future, where you can specify an AWS key or similar, which certain modules like gowitness or nuclei could use to execute on a remote VM instead of the local system.

Right now we are looking for help with this feature. If anyone has a good idea for how to implement it, or experience in that kind of setup, I am happy to work with them.

**Comment by N7WEra** [comment]
I will add Ax framwork to the mix. It’s from the same author of AXIOM, he moved to work on Ax instead of Axiom.

**Comment by DrorDvash** [comment]
Axiom / [Ax](https://github.com/attacksurge/ax) would be awesome! 

From Axiom Github:
> WARNING: Axiom Classic is now in maintenance mode. We encourage you to transition to the new [Ax Framework](https://github.com/attacksurge/ax). In the coming months, we will introduce essential quality-of-life updates to Axiom Classic before 2025. Long live Axiom!

Agree with the recommendations - I think you should consider integrating into one of them, similar to what reconFTW did.

The ability to distribute scans across multiple VPSs could make BBOT the king of all time :)

**Note regarding the provider:** AWS as alone isn’t enough, options like Linode, DigitalOcean, Hetzner, ...., is also needed.
